File: experiments/014-LiveTypingFiveRuns/cells/5-LiveTyping_5-LiveTyping+8-DesignHeuristicsInline_free/2024-1c-parcial1/20260917-031028-16596/exercise/java/src/main/java/simshotels/Floor.java
package simshotels;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class Floor {

    private ArrayList<Room> rooms;
    private HashMap<String, Integer> prices;

    // testing

    public boolean isAvailable() {
        return totalRooms() == totalRoomsAvailable();
    }

    // accessing

    public ArrayList<Room> rooms() {
        return rooms;
    }

    // accounting

    public int totalLosses() {
        int total;

        total = 0;
        for (Room room : rooms) {
            if (room.isAvailable()) {
                total = total + Collections.max(prices.values());
            }
        }

        return total;
    }

    public int totalProfits() {
        int total;

        total = 0;
        for (Room room : rooms) {
            Integer roomTotal;
            if (room.isOccupied()) {
                roomTotal = room.profitUsingIfAbsentGuestType(prices, () -> {
                    Floor.signalUnknownGuestType();
                    return null;
                });
                total = total + roomTotal;
            }
        }

        return total;
    }

    // setters

    public void setNumberOfRooms(Number aNumberOfRooms) {
        Floor.assertIsPositiveIfFalse(aNumberOfRooms, () -> Floor.signalNumberOfRoomsMustBePositive());
        Floor.assertIsIntegerIfFalse(aNumberOfRooms, () -> Floor.signalNumberOfRoomsMustBeInteger());

        initializeRoomsWith(aNumberOfRooms);
    }

    public void setPrices(HashMap<String, Integer> aPriceList) {
        if (aPriceList.isEmpty()) {
            Floor.signalNoPrices();
        }

        for (Integer price : aPriceList.values()) {
            Floor.assertIsPositiveIfFalse(price, () -> Floor.signalPriceMustBePositive());
        }

        initializeWith(aPriceList);
    }

    // totals

    public int totalRooms() {
        return rooms.size();
    }

    public int totalRoomsAvailable() {
        int count = 0;
        for (Room room : rooms) {
            if (room.isAvailable()) {
                count = count + 1;
            }
        }
        return count;
    }

    public int totalRoomsOccupied() {
        int count = 0;
        for (Room room : rooms) {
            if (room.isOccupied()) {
                count = count + 1;
            }
        }
        return count;
    }

    public int totalRoomsReserved() {
        int count = 0;
        for (Room room : rooms) {
            if (room.isReserved()) {
                count = count + 1;
            }
        }
        return count;
    }

    // initialization

    private void initializeRoomsWith(Number aNumberOfRooms) {
        int ix;

        rooms = new ArrayList<>();

        ix = 1;
        while (ix <= aNumberOfRooms.intValue()) {
            rooms.add(new Room());
            ix = ix + 1;
        }
    }

    private void initializeWith(HashMap<String, Integer> aPriceList) {
        prices = aPriceList;
    }

    // assertions (class side)

    public static void assertIsIntegerIfFalse(Number aNumber, Runnable signalsBlock) {
        if (!(aNumber instanceof Integer)) {
            signalsBlock.run();
        }
    }

    public static void assertIsPositiveIfFalse(Number aNumber, Runnable signalsBlock) {
        if (!(aNumber.doubleValue() > 0)) {
            signalsBlock.run();
        }
    }

    // error messages (class side)

    public static String numberOfRoomsMustBeIntegerErrorDescription() {
        return "Number of rooms must be integer";
    }

    public static String numberOfRoomsMustBePositiveErrorDescription() {
        return "The number of rooms must be positive";
    }

    public static String priceMustBeIntegerErrorDescription() {
        return "Price must be integer";
    }

    public static String priceMustBePositiveErrorDescription() {
        return "Price must be positive";
    }

    public static String pricesListCannotBeEmptyErrorDescription() {
        return "Prices List must not be empty";
    }

    public static String unknownGuestTypeErrorDescription() {
        return "The guest type is unknown";
    }

    // exceptions (class side)

    public static void signalNoPrices() {
        throw new RuntimeException(Floor.pricesListCannotBeEmptyErrorDescription());
    }

    public static void signalNumberOfRoomsMustBeInteger() {
        throw new RuntimeException(Floor.numberOfRoomsMustBeIntegerErrorDescription());
    }

    public static void signalNumberOfRoomsMustBePositive() {
        throw new RuntimeException(Floor.numberOfRoomsMustBePositiveErrorDescription());
    }

    public static void signalPriceMustBeInteger() {
        throw new RuntimeException(Floor.priceMustBeIntegerErrorDescription());
    }

    public static void signalPriceMustBePositive() {
        throw new RuntimeException(Floor.priceMustBePositiveErrorDescription());
    }

    public static void signalUnknownGuestType() {
        throw new RuntimeException(Floor.unknownGuestTypeErrorDescription());
    }
}

File: exercises/2024-1c-parcial1/python/src/simshotels/floor.py
from simshotels.room import Room


class Floor:

    def __init__(self):
        self._rooms = None
        self._prices = None

    # testing

    def is_available(self):
        return self.total_rooms() == self.total_rooms_available()

    # accessing

    def rooms(self):
        return self._rooms

    # accounting

    def total_losses(self):
        total = 0
        for room in self._rooms:
            if room.is_available():
                total = total + max(self._prices.values())

        return total

    def total_profits(self):
        total = 0
        for room in self._rooms:
            if room.is_occupied():
                room_total = room.profit_using_if_absent_guest_type(self._prices, lambda: Floor.signal_unknown_guest_type())
                total = total + room_total

        return total

    # setters

    def set_number_of_rooms(self, a_number_of_rooms):
        Floor.assert_is_positive_if_false(a_number_of_rooms, lambda: Floor.signal_number_of_rooms_must_be_positive())
        Floor.assert_is_integer_if_false(a_number_of_rooms, lambda: Floor.signal_number_of_rooms_must_be_integer())

        self.initialize_rooms_with(a_number_of_rooms)

    def set_prices(self, a_price_list):
        if len(a_price_list) == 0:
            Floor.signal_no_prices()

        for price in a_price_list.values():
            Floor.assert_is_positive_if_false(price, lambda: Floor.signal_price_must_be_positive())

        self.initialize_with(a_price_list)

    # totals

    def total_rooms(self):
        return len(self._rooms)

    def total_rooms_available(self):
        count = 0
        for room in self._rooms:
            if room.is_available():
                count = count + 1
        return count

    def total_rooms_occupied(self):
        count = 0
        for room in self._rooms:
            if room.is_occupied():
                count = count + 1
        return count

    def total_rooms_reserved(self):
        count = 0
        for room in self._rooms:
            if room.is_reserved():
                count = count + 1
        return count

    # initialization

    def initialize_rooms_with(self, a_number_of_rooms):
        self._rooms = []

        ix = 1
        while ix <= a_number_of_rooms:
            self._rooms.append(Room())
            ix = ix + 1

    def initialize_with(self, a_price_list):
        self._prices = a_price_list

    # assertions (class side)

    @staticmethod
    def assert_is_integer_if_false(a_number, signals_block):
        if not (isinstance(a_number, int) and not isinstance(a_number, bool)):
            signals_block()

    @staticmethod
    def assert_is_positive_if_false(a_number, signals_block):
        if not (a_number > 0):
            signals_block()

    # error messages (class side)

    @staticmethod
    def number_of_rooms_must_be_integer_error_description():
        return "Number of rooms must be integer"

    @staticmethod
    def number_of_rooms_must_be_positive_error_description():
        return "The number of rooms must be positive"

    @staticmethod
    def price_must_be_integer_error_description():
        return "Price must be integer"

    @staticmethod
    def price_must_be_positive_error_description():
        return "Price must be positive"

    @staticmethod
    def prices_list_cannot_be_empty_error_description():
        return "Prices List must not be empty"

    @staticmethod
    def unknown_guest_type_error_description():
        return "The guest type is unknown"

    # exceptions (class side)

    @staticmethod
    def signal_no_prices():
        raise RuntimeError(Floor.prices_list_cannot_be_empty_error_description())

    @staticmethod
    def signal_number_of_rooms_must_be_integer():
        raise RuntimeError(Floor.number_of_rooms_must_be_integer_error_description())

    @staticmethod
    def signal_number_of_rooms_must_be_positive():
        raise RuntimeError(Floor.number_of_rooms_must_be_positive_error_description())

    @staticmethod
    def signal_price_must_be_integer():
        raise RuntimeError(Floor.price_must_be_integer_error_description())

    @staticmethod
    def signal_price_must_be_positive():
        raise RuntimeError(Floor.price_must_be_positive_error_description())

    @staticmethod
    def signal_unknown_guest_type():
        raise RuntimeError(Floor.unknown_guest_type_error_description())
